source/slang/slang-intrinsic-expand.h

```diff
@@ -29,6 +29,26 @@ struct IntrinsicExpandContext
     Int m_argCount = 0;
     Index m_openParenCount = 0;
     CLikeSourceEmitter* m_emitter;
+
```

Comment:
This would be nice, though it assumes the name are available in IR.

Reply:
I was thinking that *if* we did this we would hypothetically move the parsing+validation step to the front-end (or maybe IR lowering?) where it could work on the AST.

Basically, we could parse the definition string using the existing logic, and translate it into a flat list of "ops" (akin to what we use for macros now):

* Raw string ops would reference a range of bytes from the original definition string
* Param reference ops would reference a parameter of the (IR-level) function by index
* Various other opcodes would be defined for the numerous special cases

During IR lowering the sequence of ops would be translated to produce zero or more operands to the `IRTargetIntrinsicDecoration`, and the original string that the front-end used would no longer be needed. The translation could just use the AST to look up parameters by name, but then write out the index of the parameter to the IR-level data.

It would be a lot of code just to remove a bit of hackery, and maybe there is a better top-down design for how to handle all of this (e.g., something like "inline assembly" and a `__switch_target` construct so that all this stuff lives inside the function body instead?).